### ISSUE TITLE
getBean() variant that accepts constructed bean?

### ISSUE THREAD

**Issue by seancorfield** [framework-one/fw1#438]
This would address the oft-raised question about how to combine DI/1 with CFML's evil ORM. A possibility would be:

    var foo = beanFactory.getBean( entityLoadByPK( id ), "fooBean" );

This would use the supplied bean as a `"fooBean"` instead of calling `construct()` to get the instance.

The constructor would be skipped -- the bean here is already constructed by the evil ORM -- so it would just do property injection.

I believe this is relatively non-intrusive: add a third optional `any instance = 0` argument to `resolveBean()` and `resolveBeanCreate()` and pass everything through, then use `instance` instead of calling `construct()` inside `cachable()`. And just modify `getBean()` to type check its arguments.

Downside, everyone pays (a little) for the conditional in `getBean()` -- for transients -- and the (slight) overhead of an additional argument passed through bean resolution.

Could use a new method instead of `getBean()` (but `injectProperties()` is already in use) to avoid that.

Another possibility is to overload the (mostly unused?) `injectProperties` to accept a bean name as the second argument and thread the needle that way?

**Comment by seancorfield** [comment]
To clarify that last thought:

    var foo = beanFactory.injectProperties( entityLoadByPK( id ), "fooBean" );

(as opposed to `injectProperties( bean, { a : "set", of : "properties" } )` so they are easily distinguishable)

**Comment by cameroncf** [comment]
Need to consider how this would work with an entityLoad() that returns as array of ORM Entities as well. Loop over that array? What about related entities that get lazy loaded inside the primary entity? 

I'd consider allowing all ORM beans to inherit some DI/1 ORM bean instead of injection since it prevents you from having to inject in so many hard to predict ways. This way, ORM entities can be smarter on the inside automatically with less work.

I know this goes against the "I" in DI, but I've found it to be way way easier.

**Comment by cameroncf** [comment]
Example for more clarity: 

https://gist.github.com/cameroncf/687c049cb79003ac0afff9b52054f528

It's helpful I think to imagine that DI/1 is REQUIRED to be secondary when using ORM. DI/1 cannot be in charge of creating or injecting ORM Entities without huge pain and complexity - in my experience it's best just to allow DI/1 to be a second class citizen inside ORM Entities. 

Let Hibernate control the ORM Entities, then only interface with DI/1 when you need something outside of ORM that lives in DI/1.

**Comment by seancorfield** [comment]
> entityLoad() that returns as array of ORM Entities

I'd expect the user to be responsible for calling `injectProperties()` on each element of that array.

> related entities that get lazy loaded

Yup, this is why I don't like ORM much: you lose far too much control over object creation and dependency management (and I value those higher).

> I'd consider allowing all ORM beans to inherit some DI/1 ORM bean

Interesting approach. What about using some of the ORM listener hooks to automate some of that (so that after a bean is loaded by the ORM, it runs some standard function that injects `getService()` for you so you don't need a base class)?

**Comment by cameroncf** [comment]
> Yup, this is why I don't like ORM much: you lose far too much control over object creation and dependency management (and I value those higher).

A different way to think about this is that both ORM and DI/1 want to be the de-facto bean factory and they both actually are in their own worlds. The problem is that they also often _both_ need to coexist in a common application and play nice while both are also allowed to be in complete control of their own little (very different) fiefdom of Object Creation.

In my mental model I tend to think of both ORM and DI/1 as being first class BeanFactories living side by side in the same app. But they are so different that trying to make them play by the same rules is a very frustrating and potentially impossible task. Let them be different and try to strictly separate what each is responsible for - minimize the overlap. My little snippet for getService() is more of a bridge between the two than anything else. 

So, in my  mind - the creation of ORM entities is entirely the responsibility of CF's ORM methods. DI/1 is not allowed to touch that for the most part. Then everything else not ORM is the responsibility of DI/1 - mainly singletons. For the most part, the two bean factories are only allowed to talk to each other in fairly strictly controlled ways -  using that getService() method, for example.

> What about using some of the ORM listener hooks to automate some of that

Certainly a possibility but I think lots of people will be using these listeners already and you'll run into problems where the framework is getting in someone's way. Plus, I've personally had a really hard time debugging problems in the listeners and don't really like to use them as a result.

> so you don't need a base class

Well, technically, you would not have to inherit from the base class. Folks could implement their own getService-like call and just turn around and call the ORMAwareBeanThing.cfc instead. It would be us to the developer how they would like to use it 1) Inherit it like FW/1 does in Application.cfc or 2) Programmatically access the same functionality.

**Comment by cameroncf** [comment]
Added a third example to the gist showing how this could be accomplished without inheritance at all, using same bean.

https://gist.github.com/cameroncf/687c049cb79003ac0afff9b52054f528

**Comment by seancorfield** [comment]
> Certainly a possibility but I think lots of people will be using these listeners already and you'll run into problems where the framework is getting in someone's way.

I wasn't suggesting the framework used it -- I was just suggesting that an application could use them (to inject `getService()` magically as each bean was loaded).

> Plus, I've personally had a really hard time debugging problems in the listeners and don't really like to use them as a result.

Well, yes, no comment. CFML's ORM implementation is horribly buggy in both engines, in different ways, and people bump into the ORM's limitations all over the place.

**Comment by cameroncf** [comment]
> I wasn't suggesting the framework used it -- I was just suggesting that an application could use them 

Perfect. We are in agreement there.

> Well, yes, no comment.

Okay.

> CFML's ORM implementation is horribly buggy in both engines, in different ways, and people bump into the ORM's limitations all over the place.

Wait, that was a comment!

But seriously, many of us love to hate the way ORM is implemented in CF but knowing that, I think it's a good idea to isolate these two engines from each other and just build a way for them to talk to each other without trying to force ORM to behave like DI/1 or even to be injectable. I prefer to let them do their own things their own ways, however messed up that is.

**Comment by seancorfield** [comment]
You're definitely persuading me that this ticket isn't needed for ORM usage. Part of me feels like it might still be useful in other situations (for example, when you have a custom factory but still want D/I for properties). The current `injectProperties()` has always felt a bit anemic to me and somewhat out of place (since it doesn't do anything that _depends_ on DI/1's machinery).

**Comment by cameroncf** [comment]
I think the most useful thing that could happen would be to define a way for someone to grab the DI/1 Bean Factory _outside of FW/1_ without having to really understand it's storage in the application scope under the covers. The only bit of my current solution that makes me feel a little icky is that I am accessing something in the app scope that should really be under FW/1's control.

```
private any function getBeanFactory() {
    return application[ request.framework_applicationKey ].factory; // this feels icky
}
```

**Comment by seancorfield** [comment]
Any suggestions for that @cameroncf ?

**Comment by cameroncf** [comment]
Well, I'm not 100% sure. Maybe I'd be able to do something like this?

```
private any function getBeanFactory() {
    return new framework.BeanFactoryFacade().getBeanFactory(); // requires FW/1 to have already been initialized
}
```

Really, I think that anything that entirely hands off the magic of where the BeanFactory really is stored under the covers by FW/1 or DI/1 is a step in the right direction. This would potentially let developers grab the BF our of thin air without having to determine how to properly pass it into anything or inject it into anything. Which is really the root problem here - ORM entities are hard to make reliably injectable.

The BeanFactoryFacade would essentially be doing `getFw1App().getBeanFactory()` internally. 

This might cause problems with subsystem beanfactories though since you might not have enough context to know which one to return.

**Comment by seancorfield** [comment]
How would that facade even know the value of `variables.framework.applicationKey`?

**Comment by seancorfield** [comment]
I think FW/1 would have to put some of this stuff in `request` scope somewhere "publicly documented" for this stuff to be possible. Then any facade could be guaranteed to "work"...

**Comment by cameroncf** [comment]
Well, the way I have solved that problem is by stashing it in the request scope so it can be picked up elsewhere. A facade can store things in the application scope but it can also be used to store things in the request scope. perhaps inside FW/1's core file it does a quick:

`BeanFactoryFacade().setBFKey()`

**Comment by cameroncf** [comment]
Something like this: https://gist.github.com/cameroncf/45c6220faa6d64ff08b6c8ee1c45217c

**Comment by seancorfield** [comment]
See #439 for discussion of out of band access of FW/1 API.

**Comment by seancorfield** [comment]
Having implemented #439 as a way to access FW/1's bean factory inside domain objects managed by the ORM, I'm closing this as "rejected". I still like the idea of being able to inject dependencies based on setters but there are a lot of caveats in this approach (aside from some mentioned above, there's also the issue of what happens if you ask DI/1 to `injectProperties( fooObject, "barBean" )`, i.e., where the prototypical bean you name does not match the type of the constructed object).
